### PR TITLE
test: consuming this library in various scenarios 

### PR DESCRIPTION
**Pull request by pkerschbaum**
## Problem

`tiny-invariant` should compile and run in various setups. Be it Node-CJS, Node-ESM, TypeScript projects with various configurations, in Browsers, etc.  
On this matter we had some troubles in the past (#144, #151) and we have a new issue (#162).  
Iterating on this topic - _package consumption in different scenarios_ - is difficult and error-prone at the moment.

## Proposed Solution

Add tests which validate that `tiny-invariant` compiles and runs in different scenarios.

This PR implements a package.json script `test:lib-consumption` which does the following:

- Start a [Verdaccio](https://verdaccio.org/) server and publish the local build of `tiny-invariant` to it.
- For each package - called "scenario" - in [test-lib-consumption/scenarios](https://github.com/pkerschbaum/tiny-invariant/tree/0f59697da7283f84442926d808430daf7a4c7205/test-lib-consumption/scenarios), run a test case which does the following: install `tiny-invariant` from the Verdaccio server and run the script `execute-scenario`.  
  The scenario can do in this script whatever it wants.

Four scenarios are implemented with this PR. Locally I added a fifth scenario for #162, and it **successfully reproduces the issue**.  
In other words, when working on a fix for #162 one can just work against that failing scenario, while making sure that the others don't break.

## Notes

- The number of files added with this PR is quite significant, however I think it is worth the assurance it gives.
- I used [`vitest`](https://vitest.dev/) to implement the `test-lib-consumption` stuff.  
  I know that we would already have `jest` for tests, problem is that I had to implement the tests with ESM (some deps need it), and the ESM+TypeScript story in jest is still very troubling. I could not get it to work while with vitest, it just worked out-of-the-box 🤓.

### CODE REVIEW

File: jest.config.js [hunk 1]
```diff
@@ -4,4 +4,5 @@ module.exports = {
   },
   moduleFileExtensions: ['ts', 'tsx', 'js', 'jsx', 'json'],
   testEnvironment: 'node',
+  rootDir: './test',
```
Comment (pkerschbaum) [inline jest.config.js:7]: Jest would execute the .spec files in the directory `test-lib-consumption` without that.

File: test-lib-consumption/main.spec.mts [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+let scenarioDirectories = fs.readdirSync(PATHS.SCENARIOS_DIRECTORY, {
+  withFileTypes: true,
+});
+scenarioDirectories = scenarioDirectories.filter((dirent) => dirent.isDirectory());
+
+for (const scenarioDirectory of scenarioDirectories) {
+  it(`scenario "${scenarioDirectory.name}"`, async () => {
+    const pathToDirectory = path.join(PATHS.SCENARIOS_DIRECTORY, scenarioDirectory.name);
+    $.cwd = pathToDirectory;
+    await $`npm install --package-lock=false`;
```
Comment (pkerschbaum) [inline test-lib-consumption/main.spec.mts:17]: `npm install` is used here (instead of yarn) because for the scenarios it does not matter - they just fetch dependencies from Verdaccio and that's it.

**Comment by alexreardon** [comment]
This looks like a fantastic change. I am so sorry for not getting back to you sooner! I have been walking through a bit of burnout. I hope to look at this thoroughly in the new year

**Comment by pkerschbaum** [comment]
No problem!  
Yes lets get back to it next year 👍

**Comment by alexreardon** [comment]
@pkerschbaum I love the goal and outcome of this change. Something I am thinking about is that this is a lot of code and complexity to add to this (little) package. I wonder if there is some way to pull what you have done out into another place (eg a package) which then `tiny-invariant` can use (reducing the complexity of this repo). What do you think?

**Comment by pkerschbaum** [comment]
@alexreardon  I was actually able to create a package [`@pkerschbaum/pkg-consumption-test`](https://www.npmjs.com/package/@pkerschbaum/pkg-consumption-test) which exposes a command to run the tests!  
Now the changes in this repository `tiny-invariant` are only the "package consumption scenarios" (and the GitHub workflow) to run the tests :)

**Comment by pkerschbaum** [comment]
I also updated the description of the PR.
